<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgres://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM stores
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [14]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name
FROM 
(SELECT *,
RANK() OVER(ORDER BY name)
FROM stores) ranked_table
WHERE rank = 4
"""

pd.read_sql_query(sql, engine)

name
0  3061 Cub Foods / Sioux City

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [23]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, btl_price, bottle_qty
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor
LIMIT 3
"""

pd.read_sql_query(sql, engine)

vendor btl_price  bottle_qty
0  Adamba Imports Int'l Inc.     $8.99          12
1      A Hardy / U.S.A. Ltd.    $28.50           3
2      A Hardy / U.S.A. Ltd.    $28.50          30

**Prompt 3 Answer**: 
> 1)  $8.99 per btl, 12 bottles.

> 2) $29.5 per btl, 6 bottles.

> 3) $28.50 per btl, 30 bottles.


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [41]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT
SUM(total)
FROM sales
WHERE EXTRACT(MONTH FROM date) = 2 and EXTRACT(YEAR FROM date) = 2015
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**: 21295350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [48]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
vendor,
COUNT(*) as total_trans,
SUM(bottle_qty) total_bottles,
SUM(total) total_revenue
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY total_trans DESC
LIMIT 3;
"""

pd.read_sql_query(sql, engine)

vendor  total_trans  total_bottles  total_revenue
0  Pernod Ricard USA/Austin Nichols        42338         403450     7880365.88
1               Bacardi U.S.A. Inc.        26441         236047     5771242.07
2                   Diageo Americas        15725         152038     3346425.42

**Prompt 5 Answer**: Pernod Ricard USA/Austin Nichols has the most transactions, 42338 in total, and they sold 403450 bottle which make them to have a 7880365.88 revenue.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [51]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost markup_price
FROM sales 
"""

pd.read_sql_query(sql, engine)

markup_price
0              $5.75
1              $5.75
2              $5.75
3              $8.58
4              $2.50
...              ...
3049908        $5.75
3049909       $10.00
3049910        $5.75
3049911        $5.75
3049912       $10.00

[3049913 rows x 1 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [58]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name,AVG((btl_price - state_btl_cost)/state_btl_cost) Percentage_Markup
FROM sales 
LEFT JOIN
stores
ON sales.store = stores.store
GROUP BY name
ORDER BY Percentage_Markup DESC
LIMIT 1;
"""

pd.read_sql_query(sql, engine)

name  percentage_markup
0  Wal-Mart 0841 / Tipton           0.525471

**Prompt 7 Answer**: Wal-Mart 0841 / Tipton has the highest percentage of markup. (0.525471)

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [69]:
# Use a SQL query to find the answer to the above prompt.

sql = """
WITH temp_table
AS
(
SELECT name,store_address, AVG((btl_price - state_btl_cost)/state_btl_cost) Percentage_Markup
FROM sales 
LEFT JOIN
stores
ON sales.store = stores.store
GROUP BY name, store_address
ORDER BY Percentage_Markup DESC
LIMIT 5
)
SELECT * FROM temp_table
ORDER BY percentage_markup DESC
"""

pd.read_sql_query(sql, engine)

name  \
0      Wal-Mart 0841 / Tipton   
1  Wal-Mart 1546 / Iowa Falls   
2  Wal-Mart 1683 / Shenandoah   
3    Wal-Mart 3394 / Atlantic   
4   Wal-Mart 2935 / Knoxville   

                                       store_address  percentage_markup  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...           0.525471  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...           0.521499  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...           0.518903  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....           0.515094  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...           0.514591

**Prompt 8 Answer**:
> Wal-Mart 0841 / Tipton has the highest percentage markup of 0.525471, their address is 1126 Highway 38 North\nTipton, IA 527720000

> Wal-Mart 2935 / Knoxville has the lowest percentage markup of 0.514591, their address is 814 W Bell Ave\nKnoxville, IA 501380000

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [106]:
sql = """
SELECT
*
FROM
(
    SELECT 
    counties.county,
    total_bottles/population avg_bottles,
    avg_bottle_size
    FROM
    (
        SELECT
        county,
        SUM(bottle_qty) total_bottles,
        SUM(bottle_qty*bottle_size)/SUM(bottle_qty) avg_bottle_size
        FROM sales
        LEFT JOIN
        products
        ON sales.item = products.item_no
        GROUP BY county
    ) temp_table
    JOIN
    counties
    ON temp_table.county = counties.county
    ORDER BY avg_bottles DESC
    LIMIT 9
) top_9_counties
ORDER BY avg_bottle_size DESC;
"""
pd.read_sql_query(sql, engine)

county  avg_bottles  avg_bottle_size
0        Kossuth           11             1119
1      Dickinson           22             1049
2    Cerro Gordo           14             1017
3           Linn           12              916
4        Johnson           13              909
5  Pottawattamie           11              908
6           Polk           15              880
7          Scott           13              815
8     Black Hawk           15              788

**Prompt 9 Answer**: Of the 9 counties who have the highest number of bottles per capita, Kossuth has the highest sells (on average) the largest bottle size

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [122]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT COUNT(*)
FROM sales
WHERE NOT btl_price * bottle_qty = CAST(total AS MONEY)
"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: There is no transaction has a value of total that is not equal to blt_price * bottle_qty.